Replace sharness-based test suite by pytest

ejoerns opened the pull request:
In the past, we faced many limitations of the sharness-based RAUC test suite and additionally, the growing size of the `rauc.t` file made it harder to keep oversight over the test cases.

With the sharness-based tests, we were often limited to having a simple program exit code check. But the reasons for exiting in RAUC are manifold.
With pytest, we can easily check for expected stdout/stderr output. Almost all tests were extended with this simple enhancement.

The new tests are now also split up into different test files, based on their test purpose.

jluebbe reviewed (commented):
Instead of replicating the `skipif` marks everywhere, they could be defined once in a helper module as suggested here: https://docs.pytest.org/en/8.2.x/how-to/skipping.html#id1

You should probably run https://github.com/astral-sh/ruff on the new code. I'd suggest a [.ruff.toml](https://docs.astral.sh/ruff/configuration/#__tabbed_1_2) with `line-length = 119` (the same our other projects).

ejoerns commented on test/meson.build:
```diff
@@ -64,16 +64,35 @@ fakerand = executable(
-)
+if pytest.found()
+  pytests = [
+    'test_cmdline',
+    'test_rauc_info',
```
`test_cmdline` is a generic test for some comment line things while the other test were meant to test the subcommand they are named after. However, this is not fully true since I have tests named `test_rauc_sign` or `test_rauc_verify`, too. Suggestions welcome.

jluebbe replied:
I'd just drop the `_rauc` part, as they all apply to RAUC.

ejoerns commented on test/conftest.py:
```diff
@@ -0,0 +1,303 @@
+    pass
+
+
+def create_system_files():
+    shutil.rmtree("images", ignore_errors=True)
+    os.mkdir("images")
+    open("images/rootfs-0", mode="w").close()
+    open("images/rootfs-1", mode="w").close()
+    open("images/appfs-0", mode="w").close()
+    open("images/appfs-1", mode="w").close()
```
Indeed. This underwent a number of changes until it ended up like this. In the current setup a fixture is what it should be. fixed.

ejoerns commented on test/conftest.py:
```diff
@@ -0,0 +1,303 @@
+meson_build = os.path.dirname(os.path.abspath(__file__)) + "/../build/"
+
+
+@pytest.fixture(autouse=True)
```
I was lazy here at first. The `monkeypatch` fixture used has a function scope and mixing scopes is not possible

jluebbe replied:
https://github.com/pytest-dev/pytest/issues/363#issuecomment-1335631998 has a good workaround.

ejoerns replied:
Yes, I came across this or a similar, too. I'll give it a try then..

ejoerns commented:
@jluebbe Thank you for the quick review!

> Instead of replicating the `skipif` marks everywhere, they could be defined once in a helper module as suggested here: https://docs.pytest.org/en/8.2.x/how-to/skipping.html#id1

Indeed. While the skipif statements were easy to copy&paste, they are probably not that easy to maintain and change and pollute the set files a bit. Fixed.

> You should probably run https://github.com/astral-sh/ruff on the new code. I'd suggest a [.ruff.toml](https://docs.astral.sh/ruff/configuration/#__tabbed_1_2) with `line-length = 119` (the same our other projects).

Ran `ruff` on it and added `ruff` to our style-check flow.

Also the other review comments should be addressed by this updated version.

jluebbe commented on test/test_cmdline.py:
```diff
@@ -0,0 +1,92 @@
+            'rauc replace-signature input',
+            'rauc replace-signature input output',
+            'rauc info'
+            ]
```
`test_rauc_excess_args` and `test_rauc_help` could use this approach as well.

jluebbe commented:
As we don't have pytest in our container, this the tests are not yet running in CI. We should add the necessary dependencies in a separate PR first.

jluebbe commented on test/conftest.py:
```diff
@@ -0,0 +1,303 @@
+        try:
+            bus.get("de.pengutronix.rauc", "/")
+            break
+        except GLib.GError:
```
Do you want to keep pydbus for other tests that would be added later? If it's just for checking if the service is started, I'd prefer to avoid the dependency.

ejoerns replied:
I've kept it first of all and have a prepared example on how to use it for dbus tests: https://github.com/ejoerns/rauc/commit/aa33a7ef08a34782c523d281898d3bf8270c5ca1

jluebbe commented on test/.ruff.toml:
```diff
@@ -0,0 +1,5 @@
+exclude = [
+    "gen_payload_tar.py"
+]
```
I've created https://github.com/rauc/rauc/pull/1406 to avoid this exclude.

ejoerns commented on test/meson.build:
```diff
@@ -64,16 +64,35 @@ fakerand = executable(
+    test(
+      'pytest_' + test_name,
+      pytest,
+      args: ['--basetemp', '/tmp/pytest-custom-root', '-vv', test_name + '.py'],
```
Right, I wanted to add a note about this. I've added it to the commit message.

jluebbe replied:
It looks like with pytest 7.3, the number of tmpdirs to keep becomes configurable:
https://docs.pytest.org/en/7.3.x/reference/reference.html#confval-tmp_path_retention_count

This is available in Debian testing, so might land in backports at some point.

jluebbe commented:
Do you also get the splice warnings (`WARNING: CPU: 1 PID: 6623 at lib/iov_iter.c:1201 iov_iter_pipe+0x31/0x40`) when running locally?

ejoerns commented:
> Do you also get the splice warnings (`WARNING: CPU: 1 PID: 6623 at lib/iov_iter.c:1201 iov_iter_pipe+0x31/0x40`) when running locally?

Yes, seems to be 9pfs-related.

ejoerns commented:
Added some missing bundle copying in convert tests that previously failed for cross (386) only (for some reason).

ejoerns commented:
Applied two minor cleanups

jluebbe commented:
Force-pushed to sort imports and to remove some redundant comments.

jluebbe commented on test/test_verify.py:
```diff
@@ -0,0 +1,114 @@
+""")
+
+
+@have_faketime
```
This should also be a module level mark.

jluebbe commented on test/test_replace_signature.py:
```diff
@@ -0,0 +1,295 @@
+    assert exitcode == 0
+    assert os.path.exists(f"{tmp_path}/bundle.sig")
+
+    out, err, exitcode = run(f"openssl asn1parse -inform DER -in {tmp_path}/bundle.sig -noout")
```
The exitcode check is missing here and in the tests below..

jluebbe commented on test/test_replace_signature.py:
```diff
@@ -0,0 +1,295 @@
+
+
+@have_openssl
+def test_replace_signature_invalid_bundle_signature_outpu(tmp_path):
```
Missing `t` at the end of the function name.

jluebbe reviewed (changes requested):
Only a few minor things left. Other improvements can wait until this is merged.

ejoerns commented on test/test_install.py:
```diff
@@ -0,0 +1,189 @@
+from helper import run
+
+
+@root
```
Moved the test case to `test_cmdline.py` now.
